File: scala/src/main/scala/Animation.scala
class Animation {

  private def onAnimate(speed:Int,c:Chamber,acc:List[String]):List[String]={
    val list=acc:+Animation.render(c)
    if(c.isEmpty)
      return list
    else
      return onAnimate(speed,c.move(speed),list)
  }

  /**
    * Animate the particles
    * @param speed speed of movement, >0
    * @param particles string representation of particles
    * @return
    */
  def animate(speed:Int, particles:String):List[String]= {
    require(speed>0 && speed<=10,"speed argument should be beween 1 and 10")
    require(particles.size>0 && particles.size<=50,"particles string should not be empty and should be less or equals 50 characters")
    onAnimate(speed, Animation.parse(particles), List[String]())
  }
}

object Animation {

  /**
    * Parse string representation of particles
    * @param particles
    * @return an instance of Chamber
    */
  def parse(particles:String):Chamber=
    particles.foldLeft[Chamber](new Chamber(0)) { (chamber, chr)=>
      chr match {
        case '.'=>new Chamber(chamber.width+1, chamber.particles)
        case 'R'=>new Chamber(chamber.width+1, chamber.particles:+Particle.R(chamber.width))
        case 'L'=>new Chamber(chamber.width+1, chamber.particles:+Particle.L(chamber.width))
        case _ =>throw new IllegalArgumentException(s"Invalid character $chr")
      }
    }

  /**
    * Render a Chamber
    * @param c
    * @return
    */
  def render(c:Chamber):String={
    val sb=new StringBuilder()
    for(i<-0 to c.width-1) {
      if(c.particles.exists(p=>p.position==i))
        sb.addOne('x')
      else
        sb.addOne('.')
    }
    sb.result()
  }
}

File: fs/Animation.Test/Animation.fs
module Animation

open System
open System.Text

type Direction =
    | Left 
    | Right

type Particle(vector:Direction, position:int32)=
    member this.Vector=vector
    member this.Position=position
    member this.move(distance:int32):Particle=match this.Vector with 
                                                | Left->Particle(this.Vector,this.Position-distance) 
                                                | Right->Particle(this.Vector,this.Position+distance)

type Chamber(width:int32, particles:Particle seq)=
    member this.Width=width
    member this.Particles=particles
    member this.moveParticles(distance:int32):Particle seq=particles 
                                                                |> Seq.map (fun x->x.move(distance)) 
                                                                |> Seq.filter (fun x->x.Position >=0 && x.Position<this.Width)
    member this.move(distance:int32):Chamber=Chamber(this.Width,this.moveParticles(distance))
    member this.empty=this.Particles |> Seq.isEmpty


let aggregate(sb:StringBuilder, i:int32, particles:Particle seq):unit=
      let particle=match particles |> Seq.tryFind(fun x->x.Position.Equals(i)) with
                        | Some p ->'x'
                        | None ->'.'
      sb.Append(particle) |> ignore

let render(c:Chamber):string=
    let sb=StringBuilder()
    let v=[0..c.Width-1] 
        |> List.iter(fun x->aggregate(sb,x,c.Particles))
    sb.ToString()

let rec animation(speed:int32,c:Chamber, acc:string list):string list=
    let l=acc @ [render(c)]
    if c.empty then
        l
    else animation(speed,c.move(speed),l)    

let parseChar(a:Chamber):char->Chamber=
    let m(c:char):Chamber=match c with
                            |'R'->Chamber(a.Width+1,a.Particles |>Seq.append [Particle(Right,a.Width)])
                            |'L'->Chamber(a.Width+1,a.Particles |>Seq.append [Particle(Left,a.Width)])
                            |'.'->Chamber(a.Width+1,a.Particles)
                            |_->a
    m

let parse(s:string):Chamber=
    s.ToCharArray() 
            |> Seq.fold(fun a c->parseChar a c) (Chamber(0, seq []))

let animate(patterns:string):int32->string list=
    let chamber=parse patterns
    let mv(speed:Int32):string list=
        animation(speed,chamber,[])
    mv

